README.md
# Prometheus Speedtest

Instrument [Speedtest.net](http://speedtest.net) tests from
[Prometheus](https://prometheus.io). Provides metrics on download\_speed,
upload\_speed, and latency. In addition, parallel installations can also be triggered to measure larger bandwidths

[![Build Status](https://travis-ci.org/jeanralphaviles/prometheus_speedtest.svg?branch=master)](https://travis-ci.org/jeanralphaviles/prometheus_speedtest/branches)
[![Docker Build Status](https://img.shields.io/docker/build/jraviles/prometheus_speedtest.svg)](https://hub.docker.com/r/jraviles/prometheus_speedtest/)
[![PyPI status](https://img.shields.io/pypi/status/prometheus_speedtest.svg)](https://pypi.python.org/pypi/prometheus_speedtest/)
[![PyPI version shields.io](https://img.shields.io/pypi/v/prometheus_speedtest.svg)](https://pypi.python.org/pypi/prometheus_speedtest/)
[![PyPI license](https://img.shields.io/pypi/l/prometheus_speedtest.svg)](https://pypi.python.org/pypi/prometheus_speedtest/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/prometheus_speedtest.svg)](https://pypi.python.org/pypi/prometheus_speedtest/)

![Grafana](https://github.com/jeanralphaviles/prometheus_speedtest/raw/master/images/grafana.png)

## Getting Started

These instructions will install and run `prometheus_speedtest` on your system.

### PyPi Package

`prometheus_speedtest` is provided as a
[PyPi package](https://pypi.org/project/prometheus_speedtest).

1. Installing

   ```shell
   pip install prometheus_speedtest
   ```

1. Running

   ```shell
   prometheus_speedtest
   ```

#### Usage

```
Instrument speedtest.net speedtests from Prometheus.
flags:

prometheus_speedtest.py:
  --address: address to listen on
    (default: '0.0.0.0')
  --excludes: speedtest server(s) to exclude - leave empty for no exclusion
    (a comma separated list)
  --port: port to listen on
    (default: '9516')
    (an integer)
  --servers: speedtest server(s) to use - leave empty for auto-selection
    (a comma separated list)
  --[no]version: show version
    (default: 'false')
  --remotes: additional speedtest trigger
    (default: '')
    (a comma separated list)
```
### Systemunit File

```
[Unit]
Description=Prometheus Speedtest Exporter
After=network.target

[Service]
User=prometheus
Group=prometheus
WorkingDirectory=/opt/speedtestprometheus
Environment="PATH=/opt/speedtestprometheus/venv/bin"
ExecStart=/opt/speedtestprometheus/venv/bin/prometheus_speedtest

[Install]
WantedBy=multi-user.target
```


### Running with Docker

`prometheus_speedtest` is also available as a [Docker](http://docker.com) image
on [Docker Hub](https://hub.docker.com/r/jraviles/prometheus_speedtest)
:whale:.

```shell
docker run --rm -d --name prometheus_speedtest -p 9516:9516/tcp jraviles/prometheus_speedtest:latest
```

You can also append extra flags when running with Docker. For example:

```shell
$ docker run --rm --name prometheus_speedtest -p 9516:9516/tcp \
    jraviles/prometheus_speedtest:latest --version
prometheus_speedtest v0.9.5
```

### Running with Kubernetes

Since you can run this from a Docker container, you can also run it in Kubernetes.

```shell
kubectl apply -f deploy/namespace.yaml
kubectl apply -f deploy/deployment.yaml
```

The Kubernetes YAML files are pre-configured to work with the
`kubernetes-pods-slow` job that comes with Prometheus, which is configured with
5m scrape times and 30s timeouts.  If you need to raise the timeout, you'll
need to change that in your Prometheus config map.

Just keep in mind, that if you increase the replica count, then Prometheus will
run a speedtest for each pod, every 5m. The same goes for if you are running
more than one replica of Prometheus, as each replica independently scrapes
targets.

### Integrating with Prometheus

`prometheus_speedtest` is best when paired with
[Prometheus](https://prometheus.io). Prometheus can be configured to perform
Speedtests on an interval and record their results.

Speedtest metrics available to query in Prometheus.

| Metric Name             | Description                           |
|------------------------ |-------------------------------------- |
| download\_speed\_bps    | Download speed (bit/s)                |
| upload\_speed\_bps      | Upload speed (bit/s)                  |
| ping\_ms                | Latency (ms)                          |
| bytes\_received         | Bytes received during test            |
| bytes\_sent             | Bytes sent during test                |
| speedtest\_device\_count| Count of devices used for measurement |

#### prometheus.yml config

Add this to your
[Prometheus config](https://prometheus.io/docs/prometheus/latest/configuration/configuration)
to start instrumenting Speedtests and recording their metrics.

```yaml
global:
  scrape_timeout: 2m

scrape_configs:
- job_name: 'speedtest'
  metrics_path: /probe
  static_configs:
  - targets:
    - localhost:9516
```

Note if you're running `prometheus` under Docker, you must link the
`prometheus` container to `prometheus_speedtest`. See the steps below for how
this can be done.

#### Trying it out

An example
[Prometheus config](https://prometheus.io/docs/prometheus/latest/configuration/configuration)
has been provided at
[example/prometheus.yml](https://github.com/jeanralphaviles/prometheus_speedtest/blob/master/example/prometheus.yml).
We'll start `prometheus` with this config.

1. Docker Network

   Create the [Docker network](https://docs.docker.com/network) that will link
   `prometheus_speedtest` and `prometheus` together.

   ```shell
   docker network create prometheus_network
   ```

1. Start Prometheus Speedtest

   ```shell
   docker run --rm -d --net prometheus_network -p 9516:9516/tcp \
      --name prometheus_speedtest jraviles/prometheus_speedtest:latest
   ```

1. Start Prometheus

   ```shell
   docker run --rm -d --net prometheus_network -p 9090:9090/tcp \
      -v $PWD/example/prometheus.yml:/etc/prometheus/prometheus.yml \
      --name prometheus prom/prometheus:latest
   ```

1. Query results

   * Visit <http://localhost:9090>

   * Wait around **45 seconds** for Prometheus to perform a Speedtest

   * Issue a query for **download\_speed\_bps**

     You should see something like this.

     ![Prometheus Query](https://github.com/jeanralphaviles/prometheus_speedtest/raw/master/images/query.png)

### Instrumenting Speedtests with cURL

Once `prometheus_speedtest` has been started, with either Docker or PyPi,
Speedtests can be instrumented with [cURL](https://curl.haxx.se).

```shell
$ curl localhost:9516/probe
# HELP speedtest_devices_total Count of devices used for Speedtest-Results (pc)
# TYPE speedtest_devices_total counter
speedtest_devices_total 3.0
# HELP download_speed_bps Download speed (bit/s)
# TYPE download_speed_bps gauge
download_speed_bps 88016694.95692767
# HELP upload_speed_bps Upload speed (bit/s)
# TYPE upload_speed_bps gauge
upload_speed_bps 3415613.277989314
# HELP ping_ms Latency (ms)
# TYPE ping_ms gauge
ping_ms 20.928
# HELP bytes_received Bytes received during test
# TYPE bytes_received gauge
bytes_received 111342756.0
# HELP bytes_sent Bytes sent during test
# TYPE bytes_sent gauge
bytes_sent 5242880.0
```

You can also visit <http://localhost:9516> in your browser to see the same
metrics.

### Default Port

Prometheus Speedtest defaults to running on port 9516; this is the allocated
port for this exporter in the
[Prometheus Default Port Allocations Guide](https://github.com/prometheus/prometheus/wiki/Default-port-allocations).

### Remotes

Prometheus Speedtest can start other speed tests in parallel using the --remotes parameter.
To do this, this Prometheus speed test must be installed on several devices.
In Prometheus, you define a main installation.
The main installation is called up with the parameter.

```shell
--remotes http://otherspeedtestdevice:9516/probe,http://anotherotherspeedtestdevice:9516/probe.
```
As soon as Prometheus queries the values of the main installation, parallel speed tests are started under the specified URLs. The main installation then returns the summed values. (An average value is determined for the ping (latency)).   

> Prometheus Speedtest supports up to 15 parallel measurements

## Getting Started (Development)

These instructions will get you a copy `prometheus_speedtest` up and running on
your local machine for development and testing purposes.

### Prerequisites

* [Python](https://www.python.org)
* [Docker](https://www.docker.com)
* [Pytest](https://pytest.org)

### Running Locally

#### Python

1. Ensure packages listed in
   [requirements.txt](https://github.com/jeanralphaviles/prometheus_speedtest/blob/master/requirements.txt)
   are installed with `pip`

   ```python
   pip3 install -r requirements.txt
   ```

1. Run `prometheus_speedtest`

   ```python
   python3 -m prometheus_speedtest.prometheus_speedtest
   ```

#### Docker

1. Building image

   ```shell
   docker build -t prometheus_speedtest:latest .
   ```

1. Running

   ```shell
   docker run --rm -d --name prometheus_speedtest -p 9516:9516/tcp prometheus_speedtest:latest
   ```

### Perform a Speedtest

```shell
curl localhost:9516/probe
```

Or visit <http://localhost:9516>

### Running Unit Tests

```shell
pytest
```

### Contributing

Pull requests are welcome. Please adhere to the
[Google Python style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md).

Please format your contributions with the
[yapf](https://github.com/google/yapf) formatter and lint your code with
[pylint](https://www.pylint.org). A
[.pylintrc](https://github.com/jeanralphaviles/prometheus_speedtest/blob/master/.pylintrc)
config has been provided.

```shell
yapf -i **/*.py
pylint **/*.py
pytype
```

## Maintenance

### Deploying to PyPi

1. Increment version number in
   [version.py](https://github.com/jeanralphaviles/prometheus_speedtest/blob/master/prometheus_speedtest/version.py)

1. Create PyPi package

   ```shell
   python3 setup.py sdist
   ```

1. Upload package to PyPi

   Ensure that [Twine](https://github.com/pypa/twine) has been installed.

   ```shell
   twine upload dist/*
   ```

### Deploying multi-architecture images to Docker Hub

1. Ensure that Docker >= 19.03 and
   [docker buildx](https://docs.docker.com/buildx/working-with-buildx/) is
   installed.

1. Build and push the new image.

   ```shell
   # Ensure you have run 'docker login'
   export DOCKER_CLI_EXPERIMENTAL=enabled
   docker buildx create --use --name my-builder
   docker buildx build --push --platform linux/amd64,linux/arm64,linux/arm/v7 \
       -t jraviles/prometheus_speedtest:latest .
   docker buildx rm my-builder
   ```

## Authors

* Jean-Ralph Aviles

## License

This product is licensed under the Apache 2.0 license. See [LICENSE](LICENSE)
file for details.

## Acknowledgments

* Matt Martz [speedtest-cli](https://github.com/sivel/speedtest-cli)
* The Prometheus team <https://prometheus.io>
* Testing in Python team <http://lists.idyll.org/listinfo/testing-in-python>
* Benjamin Staffin [python-glog](https://github.com/benley/python-glog)


CHANGELOG
# Changelog

* increase dependencies (speedtest-cli 2.1.3)
* Add remotes cli flag
  This feature allows to trigger multiple instances of this Speedtest to measure larger bandwidths from different devices


tests/prometheus_speedtest_test.py
#!/usr/bin/python3
"""Tests prometheus_speedtest.py."""

import collections
import unittest

import mock
import prometheus_client
import speedtest

from prometheus_speedtest import prometheus_speedtest


class PrometheusSpeedtestTest(unittest.TestCase):
    """Tests prometheus_speedtest.PrometheusSpeedtest."""

    _results = collections.namedtuple('Results',
                                      ['download', 'upload', 'ping'])

    @mock.patch.object(speedtest, 'Speedtest', autospec=True)
    def test_test(self, mock_speedtest):
        """Ensures correctness of PrometheusSpeedtest.test()."""
        tester = prometheus_speedtest.PrometheusSpeedtest(
            source_address='4.3.2.1', timeout=10)

        expected = PrometheusSpeedtestTest._results(10, 5, 30)
        mock_speedtest.return_value.results = expected

        self.assertEqual(tester.test(), expected)

        mock_speedtest.assert_called_once_with(source_address='4.3.2.1',
                                               timeout=10)
        mock_speedtest.return_value.get_best_server.assert_called_once_with()
        mock_speedtest.return_value.download.assert_called_once_with()
        mock_speedtest.return_value.upload.assert_called_once_with()


class SpeedtestCollectorTest(unittest.TestCase):
    """Tests prometheus_speedtest.SpeedtestCollector."""

    _results = collections.namedtuple(
        'Results',
        ['download', 'upload', 'ping', 'bytes_received', 'bytes_sent'])

    @staticmethod
    @mock.patch.object(prometheus_client.core.GaugeMetricFamily, 'add_metric')
    def test_collect(mock_metric):
        """Ensures correctness of SpeedtestCollector.collect()."""
        mock_tester = mock.create_autospec(
            prometheus_speedtest.PrometheusSpeedtest)
        collector = prometheus_speedtest.RemoteSpeedtestCollector(mock_tester)

        speedtest_results = [10, 5, 30, 100, 20]
        mock_tester.test.return_value = SpeedtestCollectorTest._results(
            *speedtest_results)

        collections.deque(collector.collect())
        mock_metric.assert_has_calls(
            [mock.call(labels=[], value=value) for value in speedtest_results])


if __name__ == '__main__':
    unittest.main()


prometheus_speedtest/prometheus_speedtest.py
#!/usr/bin/python3.7
"""Instrument speedtest.net speedtests from Prometheus."""

from typing import Optional
from typing import Sequence
from http import server
from urllib.parse import urlparse
from concurrent.futures import ThreadPoolExecutor, as_completed
import os

from absl import app
from absl import flags
from absl import logging
import prometheus_client
from prometheus_client import core
from prometheus_client.parser import text_string_to_metric_families
import speedtest
import requests
from prometheus_speedtest import version

flags.DEFINE_string('address', '0.0.0.0', 'address to listen on')
flags.DEFINE_integer('port', 9516, 'port to listen on')
flags.DEFINE_list(
    'servers', [],
    'speedtest server(s) to use - leave empty for auto-selection')
flags.DEFINE_list(
    'excludes', [],
    'speedtest server(s) to exclude - leave empty for no exclusion')
flags.DEFINE_list('remotes', [], 'additional speedtest trigger')

flags.DEFINE_boolean('version', False, 'show version')
FLAGS = flags.FLAGS


class SpeedtestData:
    """Stores all Measurement Information form speedtests"""
    def __init__(self):
        """Instantiates the SpeedtestData object."""
        self.download_speed_bps = 0
        self.upload_speed_bps = 0
        self.ping_ms = 0
        self.bytes_received = 0
        self.bytes_sent = 0
        self.count = 0

    def add(self, **kwargs):
        """Adds up all speedtest values

        Args:
            download_speed_bps: int download speed in bps
            upload_speed_bps: int upload speed in bps.
            ping_ms: int of latency to speedtest server in ms
            bytes_received: int bytes recived to determine the download speed
            bytes_sent: int bytes sent to determine the download speed
        """
        self.download_speed_bps += kwargs.get("download_speed_bps")
        self.upload_speed_bps += kwargs.get("upload_speed_bps")
        self.ping_ms += kwargs.get("ping_ms")
        self.bytes_received += kwargs.get("bytes_received")
        self.bytes_sent += kwargs.get("bytes_sent")
        self.count += 1


class PrometheusSpeedtest():
    """Enapsulates behavior performing and reporting results of speedtests."""
    def __init__(self,
                 source_address: Optional[str] = None,
                 timeout: int = 10,
                 servers: Optional[Sequence[str]] = None,
                 excludes: Optional[Sequence[str]] = None):
        """Instantiates a PrometheusSpeedtest object.

        Args:
            source_address: str - optional network address to bind to.
                e.g. 192.168.1.1.
            timeout: int - optional timeout for speedtest in seconds.
            servers: list of speedtest server ids to restrict testing to.
            servers: list of speedtest server ids to exclude testing from.
        """
        self._source_address = source_address
        self._timeout = timeout
        self._servers = servers
        self._excludes = excludes

    def test(self):
        """Performs speedtest, returns results.

        Returns:
            speedtest.SpeedtestResults object.
        """
        logging.info('Performing Speedtest...')
        client = speedtest.Speedtest(source_address=self._source_address,
                                     timeout=self._timeout)
        logging.debug(
            'Eligible servers: %s',
            client.get_servers(servers=self._servers, exclude=self._excludes))
        logging.debug('Picked server: %s', client.get_best_server())
        client.download()
        client.upload()
        logging.info('Results: %s', client.results)
        return client.results


class SpeedtestCollector():
    """Performs Speedtests when requested from Prometheus."""
    def __init__(self,
                 tester: Optional[PrometheusSpeedtest] = None,
                 servers: Optional[Sequence[str]] = None,
                 excludes: Optional[Sequence[str]] = None):
        """Instantiates a SpeedtestCollector object.

        Args:
            tester: An instantiated PrometheusSpeedtest object for testing.
            servers: servers-id to use when tester is auto-created
        """
        self._tester = tester if tester else PrometheusSpeedtest(
            servers=servers, excludes=excludes)

    def collect(self):
        """Performs a Speedtests and yields metrics.

        Return: dict with all local speedtest measurements
        """
        results = self._tester.test()
        return {
            "download_speed_bps": results.download,
            "upload_speed_bps": results.upload,
            "ping_ms": results.ping,
            "bytes_received": results.bytes_received,
            "bytes_sent": results.bytes_sent,
        }
        #
        # download_speed = core.GaugeMetricFamily('download_speed_bps',
        #                                         'Download speed (bit/s)')
        # download_speed.add_metric(labels=[], value=results.download)
        # yield download_speed
        #
        # upload_speed = core.GaugeMetricFamily('upload_speed_bps',
        #                                       'Upload speed (bit/s)')
        # upload_speed.add_metric(labels=[], value=results.upload)
        # yield upload_speed
        #
        # ping = core.GaugeMetricFamily('ping_ms', 'Latency (ms)')
        # ping.add_metric(labels=[], value=results.ping)
        # yield ping
        #
        # bytes_received = core.GaugeMetricFamily('bytes_received',
        #                                         'Bytes received during test')
        # bytes_received.add_metric(labels=[], value=results.bytes_received)
        # yield bytes_received
        #
        # bytes_sent = core.GaugeMetricFamily('bytes_sent',
        #                                     'Bytes sent during test')
        # bytes_sent.add_metric(labels=[], value=results.bytes_sent)
        # yield bytes_sent


class RemoteSpeedtestCollector():
    """
        Trigger remote Speedtests
    """
    def __init__(self,
                 tester: Optional[PrometheusSpeedtest] = None,
                 servers: Optional[Sequence[str]] = None,
                 excludes: Optional[Sequence[str]] = None,
                 remotes: Optional[Sequence[str]] = None):
        """Instantiates a SpeedtestCollector object.

        Args:
            tester: An instantiated PrometheusSpeedtest object for testing.
            servers: servers-id to use when tester is auto-created
            excludes: servers-id to exclude in tester
            remotes: urls for remote measurements
        """
        if remotes is None:
            self._remotes = []
        else:
            self._remotes = remotes
        self._tester = tester
        self._servers = servers
        self._excludes = excludes

    def collect(self):
        """
            Collects information from remote urls and local speedtest

            Yields:
                core.Metric objects.
        """
        speedtest_data = SpeedtestData()
        result = []

        with ThreadPoolExecutor(max_workers=15) as executor:
            local_speed_test = SpeedtestCollector(tester=self._tester,
                                                  servers=self._servers,
                                                  excludes=self._excludes)
            result.append(executor.submit(local_speed_test.collect))

            for url in self._remotes:
                result.append(executor.submit(remote_collector, url))

            for future in as_completed(result):
                speedtest_data.add(**future.result())

        speedtest_devices = core.CounterMetricFamily(
            'speedtest_device_count',
            'Count of devices used for Speedtest-Results (pc)')
        speedtest_devices.add_metric(labels=[], value=speedtest_data.count)
        yield speedtest_devices

        download_speed = core.GaugeMetricFamily('download_speed_bps',
                                                'Download speed (bit/s)')
        download_speed.add_metric(labels=[],
                                  value=speedtest_data.download_speed_bps)
        yield download_speed

        upload_speed = core.GaugeMetricFamily('upload_speed_bps',
                                              'Upload speed (bit/s)')
        upload_speed.add_metric(labels=[],
                                value=speedtest_data.upload_speed_bps)
        yield upload_speed

        ping = core.GaugeMetricFamily('ping_ms', 'Latency Average (ms)')
        ping.add_metric(labels=[],
                        value=speedtest_data.ping_ms / speedtest_data.count)
        yield ping

        bytes_received = core.GaugeMetricFamily('bytes_received',
                                                'Bytes received during test')
        bytes_received.add_metric(labels=[],
                                  value=speedtest_data.bytes_received)
        yield bytes_received

        bytes_sent = core.GaugeMetricFamily('bytes_sent',
                                            'Bytes sent during test')
        bytes_sent.add_metric(labels=[], value=speedtest_data.bytes_sent)
        yield bytes_sent


def remote_collector(url):
    """
        Helper function for ThreadPoolExecuter
        This function trigger the remote urls to do thier speed tests
    """
    logging.info('Fetch from URL: %s', url)
    metrics = requests.get(url).text
    result = {}
    for family in text_string_to_metric_families(metrics):
        for sample in family.samples:
            result.update({sample[0]: sample[2]})
    logging.info('Results: %s', result)
    return result


class SpeedtestMetricsHandler(server.SimpleHTTPRequestHandler,
                              prometheus_client.MetricsHandler):
    """HTTP handler extending MetricsHandler and adding status page support."""
    def __init__(self, *args, **kwargs):
        static_directory = os.path.join(os.path.dirname(__file__), 'static')
        # pylint: disable=unexpected-keyword-arg
        super().__init__(directory=static_directory, *args, **kwargs)
        # pylint: enable=unexpected-keyword-arg

    def do_GET(self):
        """Handles HTTP GET requests.

        Requests to '/probe' are handled by prometheus_client.MetricsHandler,
        other requests serve static HTML.
        """
        logging.info('%s - %s', self.requestline, self.client_address)
        path = urlparse(self.path).path
        if path == '/probe':
            prometheus_client.MetricsHandler.do_GET(self)
        else:
            server.SimpleHTTPRequestHandler.do_GET(self)


def main(argv):
    """Entry point for prometheus_speedtest.py."""
    del argv  # unused
    if FLAGS.version:
        print('prometheus_speedtest v%s' % version.VERSION)
        return

    registry = core.CollectorRegistry(auto_describe=False)
    if FLAGS.servers and set(FLAGS.excludes).issuperset(FLAGS.servers):
        logging.fatal(
            '--excludes is a superset of --includes, no viable test server is '
            'possible. Ensure --excludes does not contain all --servers.')

    # TEST_ASDF_ASDF_ASDF_ASDF_ASDF_ASDF_ASDF_ASDF_ASDF_ASDFfoobar

    registry.register(
        RemoteSpeedtestCollector(servers=FLAGS.servers,
                                 excludes=FLAGS.excludes,
                                 remotes=FLAGS.remotes))

    metrics_handler = SpeedtestMetricsHandler.factory(registry)

    http = server.ThreadingHTTPServer((FLAGS.address, FLAGS.port),
                                      metrics_handler)

    logging.info('Starting HTTP server listening on %s:%s', FLAGS.address,
                 FLAGS.port)
    http.serve_forever()


def init():
    """Initializes the prometheus_speedtest cli."""
    app.run(main)


if __name__ == '__main__':
    init()
